File: app/src/main/java/com/kelsos/mbrc/events/bus/RxBusImpl.java
package com.kelsos.mbrc.events.bus;

import com.jakewharton.rxrelay.PublishRelay;
import com.jakewharton.rxrelay.SerializedRelay;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;

import rx.Observable;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action1;

public class RxBusImpl implements RxBus {

  private SerializedRelay<Object, Object> serializedRelay = PublishRelay.create().toSerialized();
  private HashMap<Object, List<Subscription>> activeSubscriptions = new HashMap<>();

  @Override public <T> void register(Object object, final Class<T> eventClass, Action1<T> onNext) {
    //noinspection unchecked
    Subscription subscription = serializedRelay.filter(event -> event.getClass().equals(eventClass))
        .map(obj -> (T) obj)
        .subscribe(onNext);

    updateSubscriptions(object, subscription);
  }

  @Override
  public <T> void register(Object object, Class<T> eventClass, Action1<T> onNext, boolean main) {
    Subscription subscription = register(eventClass, onNext, true);
    updateSubscriptions(object, subscription);
  }

  private void updateSubscriptions(Object object, Subscription subscription) {
    List<Subscription> subscriptions = activeSubscriptions.get(object);
    if (subscriptions == null) {
      subscriptions = new LinkedList<>();
    }

    subscriptions.add(subscription);

    activeSubscriptions.put(object, subscriptions);
  }

  @Override public void unregister(Object object) {
    List<Subscription> subscriptions = activeSubscriptions.get(object);
    if (subscriptions != null) {
      Observable.from(subscriptions)
          .filter(subscription -> !subscription.isUnsubscribed())
          .subscribe(Subscription::unsubscribe);

      activeSubscriptions.remove(subscriptions);
    }
  }

  @Override public <T> Subscription register(final Class<T> eventClass, Action1<T> onNext, boolean main) {
    //noinspection unchecked
    Observable<T> observable = serializedRelay.filter(event -> event.getClass().equals(eventClass)).map(obj -> (T) obj);

    return main ? observable.observeOn(AndroidSchedulers.mainThread()).subscribe(onNext) : observable.subscribe(onNext);
  }

  @Override public void post(Object event) {
    serializedRelay.call(event);
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/events/bus/RxBusImpl.kt
package com.kelsos.mbrc.events.bus

import com.jakewharton.rxrelay2.PublishRelay
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import timber.log.Timber
import java.util.*
import javax.inject.Inject
import kotlin.reflect.KClass

class RxBusImpl @Inject
constructor() : RxBus {

  init {
    Timber.v("Injecting RxBus instance %s", this)
  }

  private val serializedRelay = PublishRelay.create<Any>().toSerialized()
  private val activeSubscriptions = HashMap<Any, MutableList<Disposable>>()

  @Suppress("UNCHECKED_CAST")
  override fun <T> register(receiver: Any, eventClass: Class<T>, onNext: (T) -> Unit) {
    //noinspection unchecked
    val subscription = serializedRelay.filter {
      it.javaClass == eventClass
    }.map { obj -> obj as T }.subscribe(onNext)

    updateSubscriptions(receiver, subscription)
  }

  override fun <T> register(receiver: Any, eventClass: Class<T>, onNext: (T) -> Unit, main: Boolean) {
    val subscription = register(eventClass, onNext, true)
    updateSubscriptions(receiver, subscription)
  }

  private fun updateSubscriptions(receiver: Any, subscription: Disposable) {
    val subscriptions: MutableList<Disposable> = activeSubscriptions[receiver] ?: LinkedList<Disposable>()
    subscriptions.add(subscription)
    activeSubscriptions.put(receiver, subscriptions)
  }

  override fun unregister(receiver: Any) {
    val subscriptions = activeSubscriptions.remove(receiver)
    if (subscriptions != null) {
      Observable.fromIterable(subscriptions).filter { !it.isDisposed }.subscribe { it.dispose() }
    }
  }

  @Suppress("UNCHECKED_CAST")
  override fun <T> register(eventClass: Class<T>, onNext: (T) -> Unit, main: Boolean): Disposable {
    //noinspection unchecked
    val observable = serializedRelay.filter { it.javaClass == eventClass }.map { obj -> obj as T }
    val scheduler = if (main) AndroidSchedulers.mainThread() else Schedulers.trampoline()
    return observable.observeOn(scheduler).subscribe(onNext)
  }

  @Suppress("UNCHECKED_CAST")
  override fun <T : Any> observe(eventClass: KClass<T>): Observable<T> {
    return serializedRelay.filter { it.javaClass == eventClass.java }.map { obj -> obj as T }
  }

  override fun post(event: Any) {
    serializedRelay.accept(event)
  }
}